Improvements to OpenStack inventory script

carsongee opened the pull request:
This adds the option to use common environment variables instead of the just the INI for common things like password and auth URL.  It creates inventory groups for each key/value in the metadata dictionary similar to the ec2.py scripts (but keeps the special group metadata tag as well for backwards compatibility), and adds the _meta host var dictionary added in 1.3 that speeds things up quite a bit. It also adds a configurable flag as to whether to prefer the public or private IP (fixed or floating) of each instance, and prints the json in a more human friendly format.

carsongee committed:
Improvements to OpenStack inventory script

mpdehaan commented:
The documentation on the new parameters is definitley a lot on the sparse side and will need to be improved if we are to consider this.

_meta is of course a great addition - perhaps these should be submitted seperately?

carsongee commented:
I've gone ahead and added documentation for the prefer_private option.  Oddly enough there was already doc for the service_type even though that parameter wasn't used at all in the script. The existing doc there is accurate though.  

Most of the changes other than that one were just refactors and additional error checking (it was raising in my environment).  It will be a bit hard to tease out the different changes since a lot of them are interdependent on the refactors.

msabramo commented:
This seems like a few different things in one PR.
1. Typo fix
2. Use `OS_*` environment variables that python-novaclient uses
3. Some new parameters and logic
4. Indented JSON output

I didn't look in depth yet but I know that we wanted 2 and I believe that someone forked nova.py internally so we could use those environment variables. So definitely looking forward to that one!

msabramo commented on plugins/inventory/nova.py:
```diff
@@ -111,12 +111,17 @@
     required: false
     default: null
-  insicure:
+  insecure:
```
AFAICT, this parameter is not used (even when spelled correctly :-)), along with a few others. See https://github.com/ansible/ansible/pull/7988

carsongee replied:
I went ahead and removed this to try and avoid the merge conflict

msabramo commented on plugins/inventory/nova.py:
```diff
@@ -160,46 +165,148 @@ def nova_load_config_file():
+    auth_url=auth_url,
+    region_name=config.get('openstack', 'region_name'),
+    project_id=project_id,
+    service_type=config.get('openstack', 'service_type'),
```
It looks like you added support for `service_type`, but removed `auth_system`? Was that intentional?

carsongee replied:
No, it looks like it got missed in the shuffle. I added it back in

msabramo commented on plugins/inventory/nova.py:
```diff
@@ -160,46 +165,148 @@ def nova_load_config_file():
-    groups = {}
-
+# Default or added list option
+if (len(sys.argv)==2 and sys.argv[1]=='--list') or len(sys.argv)==1:
```
I suspect (from looking at https://github.com/ansible/ansible/blob/devel/CODING_GUIDELINES.md#misc-preferences) that @mpdehaan would prefer spaces around `==` a la PEP8.

carsongee replied:
I fixed this and several other pep8/pylint violations

msabramo commented on plugins/inventory/nova.py:
```diff
@@ -160,46 +165,148 @@ def nova_load_config_file():
 
     # Return server list
-    print json.dumps(groups)
+    print(json.dumps(groups, sort_keys=True, indent=2))
```
Cool. I like the Python 3 compatible `print` and the nicer JSON output. :+1:

msabramo commented on plugins/inventory/nova.ini:
```diff
@@ -14,7 +14,7 @@ api_key       =
 
 # Authentication system
-auth_system =
+auth_system = keystone
```
Just curious why you set `keystone` here when it used to be blank?

msabramo commented:
So internally we have a version of `nova.py` where we also used the `OS_*` python-novaclient environment variables but my colleague did this by changing it to use exclusively the variables and he eliminated the .ini file entirely. I wonder if it's preferable to simply eliminate the .ini file; we didn't really see the point of the .ini file when the OpenStack tooling already uses the environment variables.

What say you, @mpdehaan?

msabramo committed:
nova.py: Set defaults for OpenStack settings

- auth_system
- region_name
- service_type

These are config settings that could be left out in many scenarios, but
the current code is requiring them. In particular, "service_type" is a
new one in PR #7444 so if we add that and don't set a default, then
existing .ini files won't work:

```
  File
"/Library/Frameworks/Python.framework/Versions/2.7/lib/python2.7/ConfigParser.py",
line 618, in get
    raise NoOptionError(option, section)
ConfigParser.NoOptionError: No option 'service_type' in section:
'openstack'
```

msabramo commented on plugins/inventory/nova.py:
```diff
@@ -158,48 +164,151 @@ def nova_load_config_file():
+    region_name=config.get('openstack', 'region_name'),
+    project_id=project_id,
+    auth_system=config.get('openstack', 'auth_system'),
+    service_type=config.get('openstack', 'service_type'),
```
So I tested your PR and adding `service_type` and not creating a default for it causes it to fail with an older .ini file that doesn't specify `service_type`. I submitted the following PR to fix this issue: https://github.com/carsongee/ansible/pull/1

msabramo committed:
nova.py: Support OS_AUTH_SYSTEM and OS_REGION_NAME

msabramo committed:
nova.ini: Distinguish between required and optional settings

Put them in separate sections of config to make it more clear what is
essential and what is not. Also comment out the optional settings.
And remove duplicate mention of `service_type`.

msabramo commented:
@carsongee: @jimi-c merged 2 PRs of mine that touched the same file, so this no longer merges cleanly; can you rebase? (sorry for the trouble!)

carsongee commented:
No problem. Rebased

msabramo commented:
@carsongee: @jimi-c merged another PR of mine so, sorry, looks like you have to rebase again.

carsongee commented:
Rebased again

msabramo commented:
Cool, thanks @carsongee!

Wondering what @mpdehaan and @jimi-c think now; it's evolved a little bit since the initial comments.

msabramo commented:
@jimi-c: Is this something you'd consider merging? We'd like to use the feature where it looks in the OpenStack environment variables instead of the .ini file. That's just much more convenient since we already have the environment variables.

Cc: @sontek

jimi-c commented:
@msabramo yes we're interested in this, we're just continuing to work through our backlog of issues at the moment.

msabramo commented:
@jimi-c: Cool, thanks!

msabramo commented:
@carsongee: This doesn't merge cleanly anymore and needs a little love.

carsongee commented:
Rebased and fixed a small bug I'd ran into when instances have multiple addresses.

t0mk commented:
Hi, I did another rework of the Nova inventory, I tried to cover your features and add some more, and clean up the code. If you are interested, please see:
https://github.com/ansible/ansible/pull/9378

msabramo commented:
I'm surprised this is still open. Ansible folks, can you review?
